utils/src/main/java/software/amazon/awssdk/utils/cache/NonBlocking.java

```diff
@@ -37,80 +46,204 @@
-        this(asyncThreadName, Duration.ofMinutes(1));
+        this(asyncThreadName, Duration.ofSeconds(60));
     }
 
     @SdkTestInternalApi
-    NonBlocking(String asyncThreadName, Duration asyncRefreshFrequency) {
-        this.executor = newExecutor(asyncThreadName);
-        this.asyncRefreshFrequency = asyncRefreshFrequency;
+    NonBlocking(String asyncThreadName, Duration minimumRefreshFrequency) {
+        this.asyncThreadName = asyncThreadName + "-" + THREAD_NUMBER.getAndIncrement();
+        this.minimumRefreshFrequency = minimumRefreshFrequency;
     }
 
-    private static ScheduledExecutorService newExecutor(String asyncThreadName) {
-        Validate.paramNotBlank(asyncThreadName, "asyncThreadName");
-        return new ScheduledThreadPoolExecutor(1, new ThreadFactoryBuilder().daemonThreads(true)
-                                                                            .threadNamePrefix(asyncThreadName)
-                                                                            .build());
+    @SdkTestInternalApi
+    static ThreadPoolExecutor executor() {
+        return EXECUTOR;
     }
 
     @Override
     public void initializeCachedSupplier(CachedSupplier<?> cachedSupplier) {
-        scheduleRefresh(cachedSupplier);
+        this.cachedSupplier = cachedSupplier;
     }
 
-    private void scheduleRefresh(CachedSupplier<?> cachedSupplier) {
-        executor.schedule(() -> {
-            try {
-                cachedSupplier.get();
-            } finally {
-                scheduleRefresh(cachedSupplier);
-            }
-        }, asyncRefreshFrequency.toMillis(), MILLISECONDS);
+    @Override
+    public void prefetch(Runnable valueUpdater) {
+        // Only run one async prefetch at a time.
+        if (currentlyPrefetching.compareAndSet(false, true)) {
+            tryRunBackgroundTask(valueUpdater, () -> currentlyPrefetching.set(false));
+        }
     }
 
     @Override
-    public void prefetch(Runnable valueUpdater) {
-        // Only run one async refresh at a time.
-        if (currentlyRefreshing.compareAndSet(false, true)) {
-            try {
-                executor.submit(() -> {
-                    try {
-                        valueUpdater.run();
-                    } catch (RuntimeException e) {
-                        log.warn(() -> "Exception occurred in AWS SDK background task.", e);
-                    } finally {
-                        currentlyRefreshing.set(false);
-                    }
-                });
-            } catch (Throwable e) {
-                currentlyRefreshing.set(false);
-                throw e;
-            }
+    public <T> RefreshResult<T> fetch(Supplier<RefreshResult<T>> supplier) {
+        RefreshResult<T> result = supplier.get();
+        if (result.staleTime() == null || result.prefetchTime() == null) {
+            return result;
+        }
+
+        getRefreshTime(result).ifPresent(this::schedulePrefetch);
```

Comment:
Pondering the implications of having two prefetch times - this one (jittered, used for the pre-emptive/scheduled prefetching) and the one in RefreshResult (which is the lower bound of the jittered time).

The one in RefreshResult is used by CacheSupplier on each access - for an active cred provider, that would (at prefetch time) trigger an async prefetch, thereby cancelling (and rescheduling) a subsequent prefetch task.  For those active cred providers, the scheduled prefetches are rarely going to trigger.  Leaving the scheduled pre-fetches for little-used cred-providers.

Think this is OK, no obvious implications.

Reply:
Hmmm.  This does mean that the triggered-prefetches aren't jittered.  Perhaps we narrow this down to a single, jittered prefetch value in RefreshResult so that all consumers of that are prefetching at the same time (and perhaps add a few seconds for the scheduled one, as it may get handled by the triggered one).

Reply:
We're in this deep already, we might as well keep going!

Reply:
lol.  Seems like we're So. Very.  Close. !!

Reply:
Okay, latest revision moves the jitter to the cached supplier and changed the prefetch strategy to schedule the async refresh for 1 second after the prefetch time.